core/pkg/artifacts/file_cache.go

```diff
@@ -59,11 +71,17 @@ func UserCacheDir() string {
 
 // AddFile copies a file into the cache and returns the B64MD5 cache key.
 func (c *FileCache) AddFile(path string) (string, error) {
+	c.fileSemaphore <- struct{}{}
+	defer func() { <-c.fileSemaphore }()
+
 	return addFile(c, path)
 }
 
 // AddFile computes the base-64 MD5 hash of the file and returns it. It doesn't write.
 func (c *HashOnlyCache) AddFile(path string) (string, error) {
+	c.fileSemaphore <- struct{}{}
+	defer func() { <-c.fileSemaphore }()
+
```

Comment:
It would make sense to me to push this down to `addFile`. That probably requires making the semaphore part of the interface, but I don't mind that. Entirely up to you, I don't have strong feelings here.

Reply:
I think that makes sense. I'll look at this a bit more and maybe do this in a separate PR, if that's alright!

I just realized that the test timeouts were caused by my `namedgoroutine` change. I'm going to spend my remaining time fixing that.